# Неск олько примеров использования Numpy (часть 2)

## Транспонирование массивов

Импортируем numpy:

In [1]:
import numpy as np

Создадим двумерный массив:

In [4]:
a = np.array([
    [1, 2, 3],
    [4, 5, 6]])

Метод `transpose` возвращает транспонированный массив:

In [8]:
a.transpose()

array([[1, 4],
       [2, 5],
       [3, 6]])

`T` - это короткая запись метода `transpose`:

In [5]:
a.T

array([[1, 4],
       [2, 5],
       [3, 6]])

## Сортировка массивов

Сортировка каждой строки массива:

In [9]:
data = np.array([
    [7, 1, 4],
    [8, 6, 5],
    [1, 2, 3]
])
np.sort(data)

array([[1, 4, 7],
       [5, 6, 8],
       [1, 2, 3]])

Сортировка всего массива без сохранения его формы:

In [10]:
np.sort(data, axis=None)

array([1, 1, 2, 3, 4, 5, 6, 7, 8])

Сортировка каждого столбца массива:

In [11]:
np.sort(data, axis=0)

array([[1, 1, 3],
       [7, 2, 4],
       [8, 6, 5]])

## Объединение массивов

Объединять массивы можно с помощью функции `concatenate`("сцепить"), а также с помощью более удобных для чтения функций `hstack`, `vstack` и `dstack` (для трёхмерных массивов).

In [13]:
a = np.array([
     [4, 8],
     [6, 1]
 ])
b = np.array([
     [3, 5],
     [7, 2],
 ])

np.hstack((a, b))

array([[4, 8, 3, 5],
       [6, 1, 7, 2]])

In [15]:
np.vstack((b, a))

array([[3, 5],
       [7, 2],
       [4, 8],
       [6, 1]])

Обратите внимание на две пары скобок.

Функция `concatenate` по умолчанию работает с осью 0 (в случае двумерных массивов объединяет строки исходных массивов, в случае трёхмерных - объединяет "слои"-двумерные массивы исходных массивов)

In [17]:
np.concatenate((a, b))

array([[4, 8],
       [6, 1],
       [3, 5],
       [7, 2]])

In [19]:
a_3d = np.array([
    [[1, 1],
     [1, 1]],
    [[2, 2],
     [2, 2]]
])

b_3d = np.array([
    [[3, 3],
     [3, 3]],
    [[4, 4],
     [4, 4]]
])
np.concatenate((a_3d, b_3d))

array([[[1, 1],
        [1, 1]],

       [[2, 2],
        [2, 2]],

       [[3, 3],
        [3, 3]],

       [[4, 4],
        [4, 4]]])

`axis=None` сделает массивы одномерными и объединит:

In [20]:
np.concatenate((a, b), axis=None)

array([4, 8, 6, 1, 3, 5, 7, 2])

## Упражение 1

Создайте квадратный массив (размером больше 2), отсортируйте его столбцы, получите транспонированный массив (от отсортированного), объедините эти массивы вертикально, горизонтально и без оси (выведите на экран результат для каждого действия).

# Оптимизация хранения данных
## Числовые данные

По умолчанию, целые и дробные числа в Numpy занимают 64 бита. Если чисел много, они потребуют большое количество памяти, а в некоторых случаях можно сократить это количество, используя меньше бит на одно число.

In [27]:
a = np.array([2, 2**4, 2**7 - 1, 2**7, -2**7, -2**7 - 1], dtype=np.int8)
a

array([   2,   16,  127, -128, -128,  127], dtype=int8)

В этом примере использовался целый знаковый тип, занимающий 8 бит. Как мы видим, он может хранить числа от -128 до 127, остальные числа он хранит неправильно.

In [29]:
a = np.array([2, 2**4, 2**7 - 1, 2**7, -2**7, -2**7 - 1, 2**8 -1, 2**8], dtype=np.uint8)
a

array([  2,  16, 127, 128, 128, 127, 255,   0], dtype=uint8)

Беззнаковый целый тип того же размера позволяет хранить числа от 0 до 255.

Беззнаковый тип размера *n* позволяет хранить числа от 0 до `2**n - 1`, а знаковый того же размера - от `-2**(n-1)` до `2**(n-1) - 1`.

## Упражнение 2

Напишите функцию, которая превращает список целых чисел (тип `list`) в `np.array`, занимающий наименьшую память, если известно, что данные, находящиеся в исходном списке - это:
- количество людей в каждом городе России;
- температура в градусах Цельсия, измеренная в разных хранилищах лаборатории (инкубатор для клеток млекопитающих, шкаф, холодильник, морозилка, кельвинатор -80℃)
- оценки 1 курса ФБМФ за письменный экзамен по матанализу.

## Структурированные массивы

В NumPy есть специальный вид массива, называемый массивом записей или структурированным массивом, в котором можно указать тип и, по желанию, имя для каждого столбца. Это делает сортировку и фильтрацию еще более мощной, и по ощущениям может напоминать работу с данными в Excel или других таблицах.

In [30]:
data = np.array([
    ("joe", 32, 6),
    ("mary", 15, 20),
    ("felipe", 80, 100),
    ("beyonce", 38, 9001),
], dtype=[("name", str, 10), ("age", int), ("power", int)])

data[0]

('joe', 32, 6)

In [31]:
data["name"]

array(['joe', 'mary', 'felipe', 'beyonce'], dtype='<U10')

In [32]:
data[data["power"] > 9000]["name"]

array(['beyonce'], dtype='<U10')

## Упражнение 3

Прочитайте данные из файла `Cell_culture.txt` о количестве суспензионной культуры, заказанной сотрудниками и студентами лаборатории на определённые даты, в структурированный массив. Выведите на экран:
- суммарное количество культуры, заказанной на 5 апреля;
- имена студентов и сотрудников, которые заказали культуру на 9 апреля